bindings/mqtt3/mqtt.go

```diff
@@ -195,20 +200,20 @@ func (m *MQTT) Read(ctx context.Context, handler bindings.Handler) error {
 	return nil
 }
 
-func (m *MQTT) connect(clientID string, isSubscriber bool) (mqtt.Client, error) {
+func (m *MQTT) connect(ctx context.Context, clientID string, isSubscriber bool) (mqtt.Client, error) {
 	uri, err := url.Parse(m.metadata.url)
 	if err != nil {
 		return nil, err
 	}
 	var opts *mqtt.ClientOptions
 	if isSubscriber {
-		opts = m.createSubscriberClientOptions(uri, clientID)
+		opts = m.createSubscriberClientOptions(ctx, uri, clientID)
 	} else {
 		opts = m.createClientOptions(uri, clientID)
 	}
 	client := mqtt.NewClient(opts)
 
-	ctx, cancel := context.WithTimeout(m.ctx, defaultWait)
+	ctx, cancel := context.WithTimeout(ctx, defaultWait)
```

Comment:
See above

Reply:
We do because if the parent context is cancelled we want to return straight away without trying to complete the connection. In a shutdown scenario, we want to return straight away, rather than waiting for the connection (indefinitely) or `defaultWait`.

Reply:
It does, however... Because if the context times out, `err = ctx.Err()` is non-nil, so we return an error here: `return nil, fmt.Errorf("failed to connect: %w", err)`. This causes `getProducer` not to store the created client:

```go
    p, err := m.connect(ctx, producerClientID, false)
	if err != nil {
		return nil, err
	}
	m.producer = p
```

Reply:
Is that not the exact behavior we want? The caller should expect that no client was initialized and authed if the given conext was cancelled, and the error tells them that this is the case.

Reply:
The original intent of the code was to have the connection lazily created. IMHO, a canceled message should not prevent the connection from being created, since it would unnecessarily cause the connection (which could have been established otherwise) to be attempted to be created again on the next invocation. IMHO that seems wasteful.

Reply:
I disagree that it is wasteful. I argue that is more wasteful to be creating connections, which are an expensive operation, that end up not being used. This is also going to significantly slow down program termination (10 seconds) when the root context is closed before a producer has been created.

In the interest of getting the PR merged I'll remove the context from controlling establishing connections.